daemon/device-plugin/deviceplugin.go

```diff
@@ -0,0 +1,243 @@
+package deviceplugin
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/openshift/dpu-operator/dpu-cni/pkgs/cnitypes"
+	pb "github.com/openshift/dpu-operator/tree/main/dpu-api/gen"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/types/known/emptypb"
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	VendorPluginSocketPath string = cnitypes.DaemonBaseDir + "vendor-plugin/vendor-plugin.sock"
+
+	// Device plugin settings.
+	pluginMountPath = "/var/lib/kubelet/device-plugins"
+	kubeletEndpoint = "kubelet.sock"
+	pluginEndpoint  = "sriovNet.sock"
+	resourceName    = "intel.com/ipu"
+	prefix          = "intel.com"
+)
+
+// sriovManager manages sriov networking devices
+type apfManager struct {
+	socketFile string
+	devices    map[string]pluginapi.Device // for Kubelet DP API
+	grpcServer *grpc.Server
+	pluginapi.DevicePluginServer
+	log    logr.Logger
+	client pb.DeviceServiceClient
+	conn   *grpc.ClientConn
+}
+
+type DevicePlugin interface {
+	Start() error
+}
+
+func (sm *apfManager) ListAndWatch(empty *pluginapi.Empty, stream pluginapi.DevicePlugin_ListAndWatchServer) error {
+	changed := true
+	for {
+		for id, dev := range sm.devices {
+			state := sm.GetDeviceState(id)
+			if dev.Health != state {
+				changed = true
+				dev.Health = state
+				sm.devices[id] = dev
+			}
+		}
+		if changed {
+			resp := new(pluginapi.ListAndWatchResponse)
+			for _, dev := range sm.devices {
+				resp.Devices = append(resp.Devices, &pluginapi.Device{ID: dev.ID, Health: dev.Health})
+			}
+			fmt.Printf("ListAndWatch: send devices %v\n", resp)
+			if err := stream.Send(resp); err != nil {
```

Comment:
ListAndWatch is an implemented function, I can change what's inside, but not the function headings. 

to clarify - you want the ListAndWatch to make the call to another function which will return the list of devices ? 
so instead of the devices being populated on startup, you want it routinely check?

Reply:
I have not reimplemented this yet - I would like a sync to discuss flow before changes, in case I misunderstood.

Reply:
```
func (sm *apfManager) ListAndWatch(empty *pluginapi.Empty, stream pluginapi.DevicePlugin_ListAndWatchServer) error {
	changed := true
	for {
             // call helper here that does the heavy lifting.
             // take result and pass it into stream
        }
}
```

That's all.

Reply:
done